wgpu/src/api/shader_module.rs

```diff
@@ -186,8 +186,10 @@ pub enum ShaderSource<'a> {
         shader: Cow<'a, str>,
         /// The shader stage that the shader targets. For example, `naga::ShaderStage::Vertex`
         stage: naga::ShaderStage,
-        /// Defines to unlock configured shader features.
-        defines: naga::FastHashMap<String, String>,
+        /// Key-value pairs to represent defines sent to the glsl preprocessor.
+        ///
+        /// If the same name is defined multiple times, the last value is used.
+        defines: &'a [(String, String)],
```

Comment:
Since we now clone these to make a map, I don't think there's any advantage to this being `String` over `&'a str`.

Reply:
Yeah, we should definitely take advantage to avoid allocations, if we can! 😀

Reply:
There's some question if people are going to keep their list of defines around they need `[(String, String)]`, so they'd need to remake the array every time. Probably not a big enough problem in practice, this isn't hot code.